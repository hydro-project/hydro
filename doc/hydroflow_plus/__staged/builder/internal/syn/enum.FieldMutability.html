<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../../../../hydroflow_plus/internal/syn/enum.FieldMutability.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../../../../hydroflow_plus/internal/syn/enum.FieldMutability.html">../../../../../hydroflow_plus/internal/syn/enum.FieldMutability.html</a>...</p>
    <script>location.replace("../../../../../hydroflow_plus/internal/syn/enum.FieldMutability.html" + location.search + location.hash);</script>
</body>
</html>