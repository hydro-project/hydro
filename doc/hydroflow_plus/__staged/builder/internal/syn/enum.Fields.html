<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../../../../hydroflow_plus/internal/syn/enum.Fields.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../../../../hydroflow_plus/internal/syn/enum.Fields.html">../../../../../hydroflow_plus/internal/syn/enum.Fields.html</a>...</p>
    <script>location.replace("../../../../../hydroflow_plus/internal/syn/enum.Fields.html" + location.search + location.hash);</script>
</body>
</html>