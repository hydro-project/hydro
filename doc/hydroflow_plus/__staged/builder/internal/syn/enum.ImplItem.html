<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../../../../hydroflow_plus/internal/syn/enum.ImplItem.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../../../../hydroflow_plus/internal/syn/enum.ImplItem.html">../../../../../hydroflow_plus/internal/syn/enum.ImplItem.html</a>...</p>
    <script>location.replace("../../../../../hydroflow_plus/internal/syn/enum.ImplItem.html" + location.search + location.hash);</script>
</body>
</html>