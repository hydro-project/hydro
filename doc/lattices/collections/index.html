<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Simple singleton or array collection with `cc_traits` implementations."><title>lattices::collections - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (10fa3c449 2025-04-26)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module collections</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate lattices</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">lattices</a></div><h1>Module <span>collections</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/lattices/collections.rs.html#1-1203">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Simple singleton or array collection with [<code>cc_traits</code>] implementations.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.ArrayMap.html" title="struct lattices::collections::ArrayMap">Array<wbr>Map</a></dt><dd>An array wrapper representing a fixed-size map.</dd><dt><a class="struct" href="struct.ArraySet.html" title="struct lattices::collections::ArraySet">Array<wbr>Set</a></dt><dd>An array wrapper representing a fixed-size set (modulo duplicate items).</dd><dt><a class="struct" href="struct.EmptyMap.html" title="struct lattices::collections::EmptyMap">Empty<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.EmptySet.html" title="struct lattices::collections::EmptySet">Empty<wbr>Set</a></dt><dd>A type that will always be an empty set.</dd><dt><a class="struct" href="struct.OptionMap.html" title="struct lattices::collections::OptionMap">Option<wbr>Map</a></dt><dd>A key-value entry wrapper around <code>Option&lt;(K, V)&gt;</code> representing a singleton or empty map.</dd><dt><a class="struct" href="struct.OptionSet.html" title="struct lattices::collections::OptionSet">Option<wbr>Set</a></dt><dd>A wrapper around <code>Option</code>, representing either a singleton or empty set.</dd><dt><a class="struct" href="struct.SingletonMap.html" title="struct lattices::collections::SingletonMap">Singleton<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.SingletonSet.html" title="struct lattices::collections::SingletonSet">Singleton<wbr>Set</a></dt><dd>A wrapper around an item, representing a singleton set.</dd><dt><a class="struct" href="struct.VecMap.html" title="struct lattices::collections::VecMap">VecMap</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively implemented map.</dd><dt><a class="struct" href="struct.VecSet.html" title="struct lattices::collections::VecSet">VecSet</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively-implemented set.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMap<wbr>Values</a></dt><dd>Trait for transforming the values of a map without changing the overall type of the data structure.</dd></dl></section></div></main></body></html>