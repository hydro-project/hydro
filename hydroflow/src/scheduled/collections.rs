pub struct Single<T>(pub T);

pub struct Iter<I: Iterator>(pub I);
