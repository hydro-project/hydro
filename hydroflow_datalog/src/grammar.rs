#[rust_sitter::grammar("datalog")]
#[allow(dead_code)]
pub mod datalog {
    #[rust_sitter::language]
    #[derive(Debug)]
    pub struct Program {
        pub rules: Vec<Declaration>,
    }

    #[derive(Debug, Clone)]
    pub enum Declaration {
        Input(#[rust_sitter::leaf(text = ".input")] (), Ident),
        Output(#[rust_sitter::leaf(text = ".output")] (), Ident),
        Rule(Rule),
    }

    #[derive(Debug, Clone)]
    pub struct Rule {
        pub target: Atom,
        #[rust_sitter::leaf(text = ":-")]
        _from: (),
        #[rust_sitter::repeat(non_empty = true)]
        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub sources: Vec<Atom>,
        #[rust_sitter::leaf(text = ".")]
        _dot: Option<()>,
    }

    #[derive(Debug, Clone)]
    pub struct Atom {
        pub name: Ident,
        #[rust_sitter::leaf(text = "(")]
        _l_paren: (),
        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub fields: Vec<Ident>,
        #[rust_sitter::leaf(text = ")")]
        _r_paren: (),
    }

    #[derive(Clone, PartialEq, Eq, Hash, Ord, PartialOrd, Debug)]
    pub struct Ident {
        #[rust_sitter::leaf(pattern = r"[a-zA-Z_][a-zA-Z0-9_]*", transform = |s| s.to_string())]
        pub name: String,
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }
}
