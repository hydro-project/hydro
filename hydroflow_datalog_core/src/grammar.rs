#[rust_sitter::grammar("datalog")]
#[allow(dead_code)]
pub mod datalog {
    #[rust_sitter::language]
    #[derive(Debug)]
    pub struct Program {
        pub rules: Vec<Declaration>,
    }

    #[derive(Debug, Clone)]
    pub enum Declaration {
        Input(#[rust_sitter::leaf(text = ".input")] (), Ident, RustSnippet),
        Output(
            #[rust_sitter::leaf(text = ".output")] (),
            Ident,
            RustSnippet,
        ),
        Async(
            #[rust_sitter::leaf(text = ".async")] (),
            Ident,
            /// A pipeline for data to be sent to another node, which must consume `(NodeID, Data)` pairs.
            RustSnippet,
            /// A pipeline for data received from another node, which must produce `Data` values.
            RustSnippet,
        ),
        Rule(Rule),
    }

    #[derive(Debug, Clone)]
    pub struct RustSnippet {
        #[rust_sitter::leaf(text = "`")]
        _start: (),
        #[rust_sitter::leaf(pattern = r#"[^`]*"#, transform = |s| s.to_string())]
        pub code: String,
        #[rust_sitter::leaf(text = "`")]
        _end: (),
    }

    #[derive(Debug, Clone)]
    pub struct Rule {
        pub target: TargetRelationExpr,

        pub rule_type: RuleType,

        #[rust_sitter::repeat(non_empty = true)]
        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub sources: Vec<Atom>,

        #[rust_sitter::leaf(text = ".")]
        _dot: Option<()>,
    }

    #[derive(Debug, Clone)]
    pub enum RuleType {
        Sync(#[rust_sitter::leaf(text = ":-")] ()),
        NextTick(#[rust_sitter::leaf(text = ":+")] ()),
        Async(#[rust_sitter::leaf(text = ":~")] ()),
    }

    #[derive(Debug, Clone)]
    pub enum Atom {
        Relation(
            #[rust_sitter::leaf(text = "!")] Option<()>,
            InputRelationExpr,
        ),
        Predicate(PredicateExpr),
    }

    #[derive(Debug, Clone)]
    pub struct InputRelationExpr {
        pub name: Ident,

        #[rust_sitter::leaf(text = "(")]
        _l_paren: (),

        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub fields: Vec<Ident>,

        #[rust_sitter::leaf(text = ")")]
        _r_paren: (),
    }

    #[derive(Debug, Clone)]
    pub struct TargetRelationExpr {
        pub name: Ident,

        pub at_node: Option<AtNode>,

        #[rust_sitter::leaf(text = "(")]
        _l_paren: (),

        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub fields: Vec<TargetExpr>,

        #[rust_sitter::leaf(text = ")")]
        _r_paren: (),
    }

    #[derive(Debug, Clone)]
    #[allow(clippy::manual_non_exhaustive)]
    pub struct AtNode {
        #[rust_sitter::leaf(text = "@")]
        _at: (),

        pub node: TargetExpr,
    }

    #[derive(Debug, Clone)]
    pub enum TargetExpr {
        Ident(Ident),
        Aggregation(Aggregation),
    }

    impl TargetExpr {
        pub fn ident(&self) -> &Ident {
            match self {
                TargetExpr::Ident(ident) => ident,
                TargetExpr::Aggregation(a) => &a.ident,
            }
        }
    }

    #[derive(Debug, Clone)]
    pub struct Aggregation {
        pub tpe: AggregationType,
        #[rust_sitter::leaf(text = "(")]
        _lparen: (),
        pub ident: Ident,
        #[rust_sitter::leaf(text = ")")]
        _rparen: (),
    }

    #[derive(Debug, Clone)]
    pub enum AggregationType {
        Max(#[rust_sitter::leaf(text = "max")] ()),
    }

    #[derive(Clone, PartialEq, Eq, Hash, Ord, PartialOrd, Debug)]
    pub struct Ident {
        #[rust_sitter::leaf(pattern = r"[a-zA-Z_][a-zA-Z0-9_]*", transform = |s| s.to_string())]
        pub name: String,
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }

    #[derive(Debug, Clone)]
    pub enum BoolOp {
        Lt(#[rust_sitter::leaf(text = "<")] ()),
        LtEq(#[rust_sitter::leaf(text = "<=")] ()),
        Gt(#[rust_sitter::leaf(text = ">")] ()),
        GtEq(#[rust_sitter::leaf(text = ">=")] ()),
        Eq(#[rust_sitter::leaf(text = "==")] ()),
    }

    #[derive(Debug, Clone)]
    pub struct PredicateExpr {
        #[rust_sitter::leaf(text = "(")]
        _l_brace: (),

        pub left: Ident,
        pub op: BoolOp,
        pub right: Ident,

        #[rust_sitter::leaf(text = ")")]
        _r_brace: (),
    }
}
