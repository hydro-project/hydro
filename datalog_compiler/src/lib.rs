use std::collections::{HashMap, HashSet};

use hydroflow_lang::{
    graph::flat_graph::FlatGraph,
    parse::{
        ArrowConnector, IndexInt, Indexing, NamedHfStatement, Operator, Pipeline, PipelineLink,
    },
};
use proc_macro2::{Span, TokenStream};
use quote::{quote, ToTokens};
use syn::{parse_quote, punctuated::Punctuated, token::Paren, Token};

mod grammar;

use grammar::datalog::*;

fn gen_datalog_program(literal: proc_macro2::Literal, root: TokenStream) -> syn::Stmt {
    let str_node: syn::LitStr = parse_quote!(#literal);
    let actual_str = str_node.value();
    let program: Program = grammar::datalog::parse(&actual_str).unwrap();

    let inputs = program
        .rules
        .iter()
        .filter_map(|decl| match decl {
            Declaration::Input(_, ident) => Some(ident.clone()),
            _ => None,
        })
        .collect::<Vec<_>>();

    let outputs = program
        .rules
        .iter()
        .filter_map(|decl| match decl {
            Declaration::Output(_, ident) => Some(ident.clone()),
            _ => None,
        })
        .collect::<Vec<_>>();

    let rules = program
        .rules
        .iter()
        .filter_map(|decl| match decl {
            Declaration::Rule(rule) => Some(rule.clone()),
            _ => None,
        })
        .collect::<Vec<_>>();

    let mut flat_graph = FlatGraph::default();
    let mut tee_counter = HashMap::new();

    let mut created_rules = HashSet::new();
    for decl in program.rules {
        let target_ident = match decl {
            Declaration::Input(_, ident) => ident.clone(),
            Declaration::Output(_, ident) => ident.clone(),
            Declaration::Rule(rule) => rule.target.name.clone(),
        };

        if !created_rules.contains(&target_ident) {
            created_rules.insert(target_ident.clone());
            flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Named(
                NamedHfStatement {
                    name: syn::Ident::new(&target_ident.name, Span::call_site()),
                    equals: Token![=](Span::call_site()),
                    pipeline: Pipeline::Operator(Operator {
                        path: parse_quote!(tee),
                        paren_token: Paren::default(),
                        args: Punctuated::new(),
                    }),
                },
            ));
        }
    }

    for target in inputs {
        let target_ident = syn::Ident::new(&format!("{}_recv", &target.name), Span::call_site());
        flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
            Pipeline::Link(PipelineLink {
                lhs: Box::new(Pipeline::Operator(Operator {
                    path: parse_quote!(recv_stream),
                    paren_token: Paren::default(),
                    args: vec![parse_quote!(#target_ident)]
                        .iter()
                        .cloned::<syn::Expr>()
                        .collect(),
                })),
                connector: ArrowConnector {
                    src: None,
                    arrow: Token![->](Span::call_site()),
                    dst: None,
                },
                rhs: Box::new(Pipeline::Name(syn::Ident::new(
                    &target.name,
                    Span::call_site(),
                ))),
            }),
        ));
    }

    for target in outputs {
        let tee_index = tee_counter.entry(target.name.clone()).or_insert(0);
        let my_tee_index = *tee_index;
        *tee_index += 1;

        flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
            Pipeline::Link(PipelineLink {
                lhs: Box::new(Pipeline::Name(syn::Ident::new(
                    &target.name,
                    Span::call_site(),
                ))),
                connector: ArrowConnector {
                    src: Some(Indexing {
                        bracket_token: syn::token::Bracket::default(),
                        index: IndexInt {
                            value: my_tee_index,
                            span: Span::call_site(),
                        },
                    }),
                    arrow: Token![->](Span::call_site()),
                    dst: None,
                },
                rhs: Box::new(Pipeline::Operator(Operator {
                    path: parse_quote!(for_each),
                    paren_token: Paren::default(),
                    args: vec![parse_quote!(|v| println!("{:?}", v))]
                        .iter()
                        .cloned::<syn::Expr>()
                        .collect(),
                })),
            }),
        ));
    }

    for rule in rules {
        let target = rule.target.name;
        let target_ident = syn::Ident::new(&target.name, Span::call_site());
        let sources: Vec<Target> = rule.sources.to_vec();

        // TODO(shadaj): more than two sources, nested join
        let mut identifier_to_bindings = HashMap::new();
        for (source_idx, source) in sources.iter().enumerate() {
            for (i, param) in source.fields.iter().enumerate() {
                let entry = identifier_to_bindings
                    .entry(param.clone())
                    .or_insert_with(HashMap::new);
                entry.insert(source_idx, i);
            }
        }

        let mut sorted_identifiers = identifier_to_bindings.keys().cloned().collect::<Vec<_>>();
        sorted_identifiers.sort_by_key(|ident| ident.name.clone());

        let identifiers_to_join = sorted_identifiers
            .iter()
            .filter_map(|ident| {
                let bindings = identifier_to_bindings.get(ident).unwrap();
                if bindings.len() > 1 {
                    Some((ident.clone(), bindings.clone()))
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();

        let join_node = syn::Ident::new(&format!("{}_join", target_ident), Span::call_site());

        let output_data = rule
            .target
            .fields
            .iter()
            .map(|field| {
                let bindings = identifier_to_bindings.get(field).unwrap();
                let source = bindings.keys().min().unwrap();
                let source_expr: syn::Expr = if sources.len() == 1 {
                    parse_quote!(v)
                } else {
                    let pair_idx = syn::Index::from(*source);
                    parse_quote!(kv.1.#pair_idx)
                };

                let source_col_idx = syn::Index::from(*bindings.get(source).unwrap());

                parse_quote!(#source_expr.#source_col_idx)
            })
            .collect::<Vec<syn::Expr>>();

        // TODO(shadaj): this is all a hack, will fall apart when we have more than usize types
        let source_types = sources
            .iter()
            .map(|source| {
                let col_types = source
                    .fields
                    .iter()
                    .map(|_| parse_quote!(usize))
                    .collect::<Vec<syn::Type>>();

                parse_quote!((#(#col_types),*))
            })
            .collect::<Vec<syn::Type>>();

        let key_type = identifiers_to_join
            .iter()
            .map(|(_, _)| parse_quote!(usize))
            .collect::<Vec<syn::Type>>();

        let after_join_map: syn::Expr = if sources.len() == 1 {
            parse_quote!(|v| (#(#output_data),*))
        } else {
            parse_quote!(|kv: ((#(#key_type),*), (#(#source_types),*))| (#(#output_data),*))
        };

        let after_join = Pipeline::Link(PipelineLink {
            lhs: Box::new(Pipeline::Operator(Operator {
                path: parse_quote!(map),
                paren_token: Paren::default(),
                args: vec![after_join_map].iter().cloned::<syn::Expr>().collect(),
            })),
            connector: ArrowConnector {
                src: None,
                arrow: Token![->](Span::call_site()),
                dst: None,
            },
            rhs: Box::new(Pipeline::Name(target_ident.clone())),
        });

        let join_and_map = if sources.len() == 1 {
            after_join
        } else {
            Pipeline::Link(PipelineLink {
                lhs: Box::new(Pipeline::Operator(Operator {
                    path: parse_quote!(join),
                    paren_token: Paren::default(),
                    args: Punctuated::new(),
                })),
                connector: ArrowConnector {
                    src: None,
                    arrow: Token![->](Span::call_site()),
                    dst: None,
                },
                rhs: Box::new(after_join),
            })
        };

        flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Named(
            NamedHfStatement {
                name: join_node.clone(),
                equals: Token![=](Span::call_site()),
                pipeline: join_and_map,
            },
        ));

        for (source_i, source) in sources.iter().enumerate() {
            let hash_keys: Vec<syn::Expr> = identifiers_to_join
                .iter()
                .map(|(ident, bindings)| {
                    if let Some(idx) = bindings.get(&source_i) {
                        let idx_ident = syn::Index::from(*idx);
                        parse_quote!(v.#idx_ident)
                    } else {
                        panic!("Could not find key that is being joined on: {:?}", ident);
                    }
                })
                .collect::<Vec<_>>();

            let tee_index = tee_counter.entry(source.name.name.clone()).or_insert(0);
            let my_tee_index = *tee_index;
            *tee_index += 1;

            let source_data_types = source
                .fields
                .iter()
                .map(|_| parse_quote!(usize))
                .collect::<Vec<syn::Type>>();

            flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
                Pipeline::Link(PipelineLink {
                    lhs: Box::new(Pipeline::Name(syn::Ident::new(
                        &source.name.name,
                        Span::call_site(),
                    ))),
                    connector: ArrowConnector {
                        src: Some(Indexing {
                            bracket_token: syn::token::Bracket::default(),
                            index: IndexInt {
                                value: my_tee_index,
                                span: Span::call_site(),
                            }
                        }),
                        arrow: Token![->](Span::call_site()),
                        dst: None,
                    },
                    rhs: Box::new(if sources.len() == 1 {
                        Pipeline::Name(join_node.clone())
                    } else {
                        Pipeline::Link(PipelineLink {
                            lhs: Box::new(Pipeline::Operator(Operator {
                                path: parse_quote!(map),
                                paren_token: Paren::default(),
                                args: vec![parse_quote!(|v: (#(#source_data_types),*)| ((#(#hash_keys),*), v))]
                                    .iter()
                                    .cloned::<syn::Expr>()
                                    .collect(),
                            })),
                            connector: ArrowConnector {
                                src: None,
                                arrow: Token![->](Span::call_site()),
                                dst: Some(Indexing {
                                    bracket_token: syn::token::Bracket::default(),
                                    index: IndexInt {
                                        value: source_i,
                                        span: Span::call_site(),
                                    },
                                }),
                            },
                            rhs: Box::new(Pipeline::Name(join_node.clone())),
                        })
                    }),
                }),
            ));
        }
    }

    println!("{}", flat_graph.surface_syntax_string());

    let code_tokens = flat_graph
        .into_partitioned_graph()
        .expect("failed to partition")
        .as_code(root);

    syn::parse_quote!({
        #code_tokens
    })
}

#[proc_macro]
pub fn datalog(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let item = proc_macro2::TokenStream::from(item);
    let literal: proc_macro2::Literal = syn::parse_quote! {
        #item
    };

    let hydroflow_crate = proc_macro_crate::crate_name("hydroflow")
        .expect("hydroflow should be present in `Cargo.toml`");
    let root = match hydroflow_crate {
        proc_macro_crate::FoundCrate::Itself => quote! { hydroflow },
        proc_macro_crate::FoundCrate::Name(name) => {
            let ident = syn::Ident::new(&name, Span::call_site());
            quote! { #ident }
        }
    };

    proc_macro::TokenStream::from(gen_datalog_program(literal, root).to_token_stream())
}

#[cfg(test)]
mod tests {
    use std::fs::File;
    use std::io::{Read, Write};
    use std::process::Command;

    use quote::ToTokens;
    use syn::parse_quote;
    use tempfile::tempdir;

    use super::gen_datalog_program;

    fn rustfmt_code(code: &str) -> String {
        let dir = tempdir().unwrap();
        let file_path = dir.path().join("temp.rs");
        let mut file = File::create(file_path.clone()).unwrap();

        writeln!(file, "{}", code).unwrap();
        drop(file);

        Command::new("rustfmt")
            .arg(file_path.to_str().unwrap())
            .spawn()
            .unwrap()
            .wait()
            .unwrap();

        let mut file = File::open(file_path).unwrap();
        let mut data = String::new();
        file.read_to_string(&mut data).unwrap();
        drop(file);
        dir.close().unwrap();
        data
    }

    #[test]
    fn minimal_program() {
        let out = &gen_datalog_program(
            parse_quote!(
                r#"
                .input in
                .output out

                out(y, x) :- in(x, y).
                "#
            ),
            quote::quote! { hydroflow },
        );

        let wrapped: syn::Item = parse_quote! {
            fn main() {
                #out
            }
        };

        insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    }

    // non-deterministic codegen
    // #[test]
    // fn join_with_self() {
    //     let out = &gen_datalog_program(
    //         parse_quote!(
    //             r#"
    //             .input in
    //             .output out

    //             out(x, y) :- in(x, y), in(y, x).
    //             "#
    //         ),
    //         quote::quote! { hydroflow },
    //     );

    //     let wrapped: syn::Item = parse_quote! {
    //         fn main() {
    //             #out
    //         }
    //     };

    //     insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    // }

    #[test]
    fn join_with_other() {
        let out = &gen_datalog_program(
            parse_quote!(
                r#"
                .input in1
                .input in2
                .output out

                out(x, y) :- in1(x, y), in2(y, x).
                "#
            ),
            quote::quote! { hydroflow },
        );

        let wrapped: syn::Item = parse_quote! {
            fn main() {
                #out
            }
        };

        insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    }
}
