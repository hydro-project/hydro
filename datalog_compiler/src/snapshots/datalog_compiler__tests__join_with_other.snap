---
source: datalog_compiler/src/lib.rs
expression: rustfmt_code(&wrapped.to_token_stream().to_string())
---
fn main() {
    {
        {
            use hydroflow::tl;
            let mut df = hydroflow :: scheduled :: graph :: Hydroflow :: new_with_graph ("{\"nodes\":[{\"value\":null,\"version\":0},{\"value\":\"op_1v1: recv_stream (in1_recv)\",\"version\":1},{\"value\":\"op_2v1: recv_stream (in2_recv)\",\"version\":1},{\"value\":\"op_3v1: for_each (| v | println ! (\\\"{:?}\\\" , v))\",\"version\":1},{\"value\":\"op_4v1: join ()\",\"version\":1},{\"value\":\"op_5v1: map (| (k , (v1 , v2)) | v1)\",\"version\":1},{\"value\":\"op_6v1: map (| v | ((v . 0 , v . 1) , v))\",\"version\":1},{\"value\":\"op_7v1: map (| v | ((v . 1 , v . 0) , v))\",\"version\":1}],\"edges\":[{\"value\":null,\"version\":0},{\"value\":[{\"idx\":6,\"version\":1}],\"version\":1},{\"value\":[{\"idx\":7,\"version\":1}],\"version\":1},{\"value\":null,\"version\":0},{\"value\":[{\"idx\":5,\"version\":1}],\"version\":1},{\"value\":[{\"idx\":3,\"version\":1}],\"version\":1},{\"value\":[{\"idx\":4,\"version\":1}],\"version\":1},{\"value\":[{\"idx\":4,\"version\":1}],\"version\":1}],\"handoffs\":[{\"value\":null,\"version\":0}],\"subgraph_nodes\":[{\"value\":null,\"version\":0},{\"value\":[{\"idx\":1,\"version\":1},{\"idx\":2,\"version\":1},{\"idx\":6,\"version\":1},{\"idx\":7,\"version\":1},{\"idx\":4,\"version\":1},{\"idx\":5,\"version\":1},{\"idx\":3,\"version\":1}],\"version\":1}],\"subgraph_stratum\":[{\"value\":null,\"version\":0},{\"value\":0,\"version\":1}]}\n") ;
            let mut in1_recv = in1_recv;
            let mut in2_recv = in2_recv;
            let mut sg_1v1_node_4v1_joindata = Default::default();
            df.add_subgraph_stratified(
                "Subgraph GraphSubgraphId(1v1)",
                0,
                tl!(),
                tl!(),
                move |context, tl!(), tl!()| {
                    let op_1v1 = std::iter::from_fn(|| {
                        match in1_recv
                            .poll_recv(&mut std::task::Context::from_waker(&mut context.waker()))
                        {
                            std::task::Poll::Ready(maybe) => maybe,
                            std::task::Poll::Pending => None,
                        }
                    });
                    let op_2v1 = std::iter::from_fn(|| {
                        match in2_recv
                            .poll_recv(&mut std::task::Context::from_waker(&mut context.waker()))
                        {
                            std::task::Poll::Ready(maybe) => maybe,
                            std::task::Poll::Pending => None,
                        }
                    });
                    let op_6v1 = op_1v1.map(|v| ((v.0, v.1), v));
                    let op_7v1 = op_2v1.map(|v| ((v.1, v.0), v));
                    let op_4v1 = hydroflow::compiled::pull::SymmetricHashJoin::new(
                        op_6v1,
                        op_7v1,
                        &mut sg_1v1_node_4v1_joindata,
                    );
                    let op_5v1 = op_4v1.map(|(k, (v1, v2))| v1);
                    let op_3v1 =
                        hydroflow::compiled::for_each::ForEach::new(|v| println!("{:?}", v));
                    hydroflow::compiled::pivot::Pivot::new(op_5v1, op_3v1).run();
                },
            );
            df
        }
    }
}

